### PR TITLE
feat(txn-table): transaction table

### LINKED ISSUE #28: [DappKit] Transaction table
refers https://eth.blockscout.com/txs

<sub><a href="https://huly.app/guest/geist?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzQyZWRhOGQyZWY0Y2JjYzQ2ODVkZDYiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctZ2l0aHViZGVidWdnLWdlaXN0LTY3NDAzNjhlLTU0NTgxY2VhOWEtYmNhZDgwIn0.mLo-La6ObaLD6axBHFSdfhHSjiqE-JP1VQvDLeZPTEc">Huly&reg;: <b>DAPPK-28</b></a></sub>

### PR DESCRIPTION
**Pull request by metalboyrick**
Fixes #28

Milestones To-do:

- [x] Create basic table UI and review if if it is desired
- [x] Create version with Integrated Blockscout API
- [ ] Add unit tests (if applicable)

<sub><a href="https://huly.app/guest/geist?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzQ4OTA0YmQyZWY0Y2JjYzQ2YzNhY2EiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctZ2l0aHViZGVidWdnLWdlaXN0LTY3NDAzNjhlLTU0NTgxY2VhOWEtYmNhZDgwIn0.Zq8-jOuwexWecbdUbOoPmLknsUj184uOQ_a1SBra_Vc">Huly&reg;: <b>DAPPK-38</b></a></sub>

### CODE REVIEW

File: apps/storybook/src/lib/blockscout/fixture.ts [hunk 1]
```diff
@@ -434,3 +434,5 @@ export const TXN_VITALIK_TRANSFER = {
 	block_number: 21067426,
 	has_error_in_internal_transactions: false,
 };
+
+export const TXN_LIST = [TXN_VITALIK_DEPOSIT, TXN_VITALIK_TRANSFER];
```
Comment (debuggingfuture) [inline apps/storybook/src/lib/blockscout/fixture.ts:438]: @metalboyrick can help to add more fixtures? 
picking
- erc20 transfer

generating  createTxnFixture()
- 100 records, for perf loading test

File: apps/storybook/src/stories/transactions/TransactionTable.tsx [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+			sortConfig &&
+			sortConfig.key === key &&
+			sortConfig.direction === "asc"
+		) {
+			direction = "desc";
+		}
+		setSortConfig({ key, direction });
+	};
+
+	return (
```
Comment (debuggingfuture) [inline apps/storybook/src/stories/transactions/TransactionTable.tsx:83]: https://ui.shadcn.com/docs/components/data-table

we could separate two components DataTable, Transaction Table in the sense that header info are decoupled?

Looking at `AttestationsTable`  see if we could share same underlying DataTable for both use cases?

File: apps/storybook/src/stories/transactions/TransactionTable.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	const currentTransactions = sortedTransactions.slice(
+		indexOfFirstItem,
+		indexOfLastItem,
+	);
+
+	const totalPages = Math.ceil(transactions.length / itemsPerPage);
+
+	const requestSort = (key: keyof TransactionMeta) => {
+		let direction: "asc" | "desc" = "asc";
+		if (
```
Comment (debuggingfuture) [inline apps/storybook/src/stories/transactions/TransactionTable.tsx:73]: if it's a toggle, we could use direction as `isAsc`  flag and just set isAsc = !isAsc?

separate flag for isSorted if we don't always want to sort

**Review by debuggingfuture — CHANGES_REQUESTED** [review]

File: apps/storybook/src/lib/blockscout/fixture.ts (hunk 1 above)
Comment (metalboyrick) [reply apps/storybook/src/lib/blockscout/fixture.ts:438]: can ser no problem

File: apps/storybook/src/stories/transactions/TransactionTable.tsx (hunk 2 above)
Comment (metalboyrick) [reply apps/storybook/src/stories/transactions/TransactionTable.tsx:83]: we can no problem, i just whipped a quick one with v0, better unify and use the data table then

File: apps/storybook/src/stories/transactions/TransactionTable.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import { DataTable } from "@/components/ui/data-table";
+import { TransactionMeta } from "@/lib/domain/transaction/transaction";
+import { getShortAddress } from "@/utils/address";
+import { getShortHex } from "@/utils/hex";
+import { ColumnDef } from "@tanstack/react-table";
+import { ArrowUpDown, ExternalLink } from "lucide-react";
+import { formatEther } from "viem";
+
+export const txnTypeMap: Record<string, string> = {
+	contract_call: "Contract Call",
```
Comment (debuggingfuture) [inline apps/storybook/src/stories/transactions/TransactionTable.tsx:11]: duplicated with `findDisplayedTxType` in `asTransactionMeta`
apps/storybook/src/lib/blockscout/api.ts

File: apps/storybook/src/stories/transactions/TransactionTable.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+export const columns: ColumnDef<TransactionMeta>[] = [
+	{
+		accessorKey: "hash",
+		header: "Txn Hash",
+		cell: ({ row }) => {
+			const txnHash = row.getValue<string>("hash") as `0x${string}`;
+
+			return (
+				<a
+					href={`https://eth.blockscout.com/tx/${txnHash}`}
```
Comment (debuggingfuture) [inline apps/storybook/src/stories/transactions/TransactionTable.tsx:33]: use apps/storybook/src/lib/explorer/url.ts for chain based url

**Review by debuggingfuture — CHANGES_REQUESTED** [review]
good stuff we just need to ensure extract common utils between TokenInfo

**Review by debuggingfuture — CHANGES_REQUESTED** [review]

File: apps/storybook/src/stories/transactions/TransactionTable.tsx (hunk 4 above)
Comment (metalboyrick) [reply apps/storybook/src/stories/transactions/TransactionTable.tsx:11]: hmm doesnt seem so, `findDisplayedTxType` is re-aliasing of the naming, while `txnTypeMap` kinda converts the `snake_case` name to a plain language.


```ts
export const findDisplayedTxType = (transaction_types: any[]): string => {
	if (transaction_types.includes("contract_call")) {
		return "contract_call";
	}
	if (transaction_types.includes("coin_transfer")) {
		return "coin_transfer";
	}
	return "native_transfer";
};
```
```ts
export const txnTypeMap: Record<string, string> = {
  contract_call: "Contract Call",
  native_transfer: "Native Token Transfer",
  token_transfer: "Token Transfer",
  coin_transfer: "Token Transfer",
};
```

Or would you prefer to  render the snake_case in the table?

File: apps/storybook/src/stories/transactions/TransactionTable.tsx (hunk 5 above)
Comment (metalboyrick) [reply apps/storybook/src/stories/transactions/TransactionTable.tsx]: Upgraded to be able to support different chains

File: apps/storybook/src/stories/transactions/TransactionTable.stories.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	parameters: {
+		layout: "centered",
+	},
+	decorators: [withWagmiProvider()],
+} satisfies Meta<typeof TransactionTable>;
+
+export default meta;
+type Story = StoryObj<typeof meta>;
+
+export const TransactionTableWithoutFetching: Story = {
```
[the remark this answers is not in the document]
Comment (metalboyrick) [reply apps/storybook/src/stories/transactions/TransactionTable.stories.ts:24]: IMO `TransactionTable` for presentation and `TransactionTableWithDetails` for fetching, the details will come from the existing block explorers

**Comment by metalboyrick** [comment]
Demo video: https://www.loom.com/share/f24b85a30e7b4c2cbe7459165dee07e9?sid=5b38c48f-128e-41b6-b138-b9178ac25063

File: apps/storybook/src/stories/transactions/TransactionTable.stories.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (debuggingfuture) [reply apps/storybook/src/stories/transactions/TransactionTable.stories.ts]: i know where you coming from, while I wonder if we make it easy to be "lazy"

lazy=>use higher level component, "smart" under the hood
lazy=>shorter name
explicit=>more control=>longer name

no conclusion yet we could bring discussion to this
https://huly.app/workbench/geist/document/implementation-guide-67403d3146d1633c4ef22c52

File: apps/storybook/src/stories/transactions/TransactionTable.tsx (hunk 4 above)
Comment (debuggingfuture) [reply apps/storybook/src/stories/transactions/TransactionTable.tsx:17]: oh my bad you're right, I quoted the wrong code
haven't even merge that
Let's use yours first I can follow up

**Pull request merged**